diesel_codegen/src/schema_inference.rs

```diff
@@ -11,19 +11,79 @@ pub fn derive_infer_schema(input: syn::MacroInput) -> quote::Tokens {
                with your invocation of `infer_schema!");
     }
 
-    let options = input.attrs.iter().find(|a| a.name() == "options").map(|a| &a.value);
-    let options = match options {
-        Some(&syn::MetaItem::List(_, ref options)) => options,
-        _ => bug(),
-    };
-    let database_url = options.iter().find(|a| a.name() == "database_url")
-        .map(|a| str_value_of_meta_item(a, "database_url"))
-        .unwrap_or_else(|| bug());
+    let options = get_options_from_input(&input.attrs, bug);
+    let database_url = get_option(&options, "database_url", bug);
 
     let table_names = load_table_names(&database_url).unwrap();
     let schema_inferences = table_names.into_iter().map(|table_name| {
         quote!(infer_table_from_schema!(#database_url, #table_name);)
-    }).collect::<Vec<_>>();
+    });
 
     quote!(#(schema_inferences)*)
 }
+
+pub fn derive_infer_table_from_schema(input: syn::MacroInput) -> quote::Tokens {
+    fn bug() -> ! {
+        panic!("This is a bug. Please open a Github issue \
+               with your invocation of `infer_table_from_schema!");
+    }
+
+    let options = get_options_from_input(&input.attrs, bug);
+    let database_url = get_option(&options, "database_url", bug);
+    let table_name = get_option(&options, "table_name", bug);
+
+    let connection = establish_connection(&database_url).unwrap();
+    let data = get_table_data(&connection, &table_name).unwrap();
+    let primary_keys = get_primary_keys(&connection, &table_name).unwrap()
```

Comment:
I'm fairly certain you can omit the `&` in `&table_name` as it's already a `&str`. (Same for `&options` and `&database_url` above.)